### PR TITLE
fix(jest): The `mock` function `factory` argument is a callback

### PR DESCRIPTION
**Pull request by ExE-Boss**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:
  - <https://jestjs.io/docs/en/jest-object#jestmockmodulename-factory-options>
  - <https://jestjs.io/docs/en/jest-object#jestdomockmodulename-factory-options>
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

---

This applies to **Jest 16** as well, all the way back to **Jest 11**, which introduced `jest.mock(…)` and `jest.doMock(…)`.

I also cleaned up `tslint.json` by removing rules that are no longer being violated.

### CODE REVIEW

File: types/jest/index.d.ts [hunk 1]
```diff
@@ -1,6 +1,6 @@
 // Type definitions for Jest 24.0
-// Project: https://jestjs.io
-// Definitions by: Asana <https://asana.com>
+// Project: https://jestjs.io/
+// Definitions by: Asana (https://asana.com)
```
Comment (Favna) [inline types/jest/index.d.ts]: Why did you change this? Please revert it...

[the remark this answers is not in the document]
Comment (ExE-Boss) [reply types/jest/index.d.ts]: Because [the&nbsp;`dt‑header` rule](https://github.com/microsoft/dtslint/blob/master/docs/dt-header.md) told me&nbsp;to.

[the remark this answers is not in the document]
Comment (Favna) [reply types/jest/index.d.ts]: There is not a single mention of listing it as `Definitions by: My Name (My not github URL)` on there. It all says `Definitions by: My Name <https://github.com/myname>`. The only thing it says is that it is ***preferred*** to list a GitHub URL but preferences are never rules and it is not your place to correct the URL someone else put in the file. And even then it should still be `<https://asana.com>`.

[the remark this answers is not in the document]
Comment (Favna) [reply types/jest/index.d.ts]: Oh you mean when you ran `npm run lint jest`? If so I guess they changed that check since I PR'd for Jest because it didn't say anything about it for me. Alright I guess.

**Review by jwbay — APPROVED** [review]
